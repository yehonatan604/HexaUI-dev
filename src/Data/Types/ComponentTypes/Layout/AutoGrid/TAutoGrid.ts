import { TComponent, ComponentOptions } from "../../TComponent";

export type TAutoGrid = TComponent<"div"> & {
    options?: AutoGridOptions;
    rowsArr: React.ReactNode[];
    cols: number;
}

export type AutoGridOptions = ComponentOptions & {
    borderVariant?: string;
}
