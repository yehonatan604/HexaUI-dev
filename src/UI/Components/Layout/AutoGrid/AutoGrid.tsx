import { TAutoGrid } from "./types/TAutoGrid";

const AutoGrid = (props: TAutoGrid) => {
  const { rowsArr, cols, options, ...componentProps } = props;

  // Options
  const showBorder = options?.border?.show || true;
  const defaultBorderColor = "border-dtandard-d dark:border-standard-l";
  const borderColor = options?.borderVariant || defaultBorderColor;

  // JSX
  return (
    <div
      role="grid"
      {...componentProps}
      style={{
        display: "inline-grid",
        gridTemplateColumns: `repeat(${cols}, 1fr)`,
      }}
    >
      {rowsArr.map((row, index) => (
        <div key={index} className={`${showBorder && "border"} border-${borderColor}`}>
          {row}
        </div>
      ))}
    </div>
  );
};

export default AutoGrid;
