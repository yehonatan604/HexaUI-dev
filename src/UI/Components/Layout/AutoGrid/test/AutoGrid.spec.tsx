import { render, screen } from "@testing-library/react";
import AutoGrid from "../AutoGrid"; // Adjust the import path
import { describe, it, expect } from "vitest";

describe("AutoGrid Component", () => {
  const rowsArr = [
    <div key="1">Row 1</div>,
    <div key="2">Row 2</div>,
    <div key="3">Row 3</div>,
    <div key="4">Row 4</div>,
  ];

  it("should render the grid with the correct number of columns", () => {
    render(<AutoGrid rowsArr={rowsArr} cols={2} />);

    // Check if the grid is rendered with the correct column structure
    const gridElement = screen.getByRole("grid");
    expect(gridElement).toBeInTheDocument();

    // Check if it has the correct grid style for two columns
    expect(gridElement).toHaveStyle({
      display: "inline-grid",
      gridTemplateColumns: "repeat(2, 1fr)",
    });
  });

  it("should render the correct number of rows", () => {
    render(<AutoGrid rowsArr={rowsArr} cols={2} />);

    // Check if the correct number of row items are rendered
    const rowElements = screen.getAllByText(/Row/);
    expect(rowElements).toHaveLength(rowsArr.length);
  });

  it("should apply default border styles when no borderVariant is provided", () => {
    render(<AutoGrid rowsArr={rowsArr} cols={2} />);

    // Check if each row has the default border class applied
    const rowElements = screen.getAllByText(/Row/);
    rowElements.forEach((row) => {
      expect(row.parentElement).toHaveClass("border");
    });
  });

  it("should apply custom border styles when borderVariant is provided", () => {
    render(
      <AutoGrid rowsArr={rowsArr} cols={2} options={{ borderVariant: "red-500" }} />
    );

    // Check if each row has the custom border class applied
    const rowElements = screen.getAllByText(/Row/);
    rowElements.forEach((row) => {
      expect(row.parentElement).toHaveClass("border", "border-red-500");
    });
  });

  it("should render and match the snapshot", () => {
    const { asFragment } = render(<AutoGrid rowsArr={rowsArr} cols={2} />);

    // Snapshot the AutoGrid structure
    expect(asFragment()).toMatchSnapshot();
  });

  it("should render with custom border styles and match the snapshot", () => {
    const { asFragment } = render(
      <AutoGrid rowsArr={rowsArr} cols={2} options={{ borderVariant: "green-500" }} />
    );

    // Snapshot the AutoGrid structure with custom border
    expect(asFragment()).toMatchSnapshot();
  });
});
