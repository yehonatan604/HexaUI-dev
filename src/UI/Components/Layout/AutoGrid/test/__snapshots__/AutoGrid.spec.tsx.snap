// Vitest Snapshot v1

exports[`AutoGrid Component > should render and match the snapshot 1`] = `
<DocumentFragment>
  <div
    role="grid"
    style="display: inline-grid; grid-template-columns: repeat(2, 1fr);"
  >
    <div
      class="border border-border-dtandard-d dark:border-standard-l"
    >
      <div>
        Row 1
      </div>
    </div>
    <div
      class="border border-border-dtandard-d dark:border-standard-l"
    >
      <div>
        Row 2
      </div>
    </div>
    <div
      class="border border-border-dtandard-d dark:border-standard-l"
    >
      <div>
        Row 3
      </div>
    </div>
    <div
      class="border border-border-dtandard-d dark:border-standard-l"
    >
      <div>
        Row 4
      </div>
    </div>
  </div>
</DocumentFragment>
`;

exports[`AutoGrid Component > should render with custom border styles and match the snapshot 1`] = `
<DocumentFragment>
  <div
    role="grid"
    style="display: inline-grid; grid-template-columns: repeat(2, 1fr);"
  >
    <div
      class="border border-green-500"
    >
      <div>
        Row 1
      </div>
    </div>
    <div
      class="border border-green-500"
    >
      <div>
        Row 2
      </div>
    </div>
    <div
      class="border border-green-500"
    >
      <div>
        Row 3
      </div>
    </div>
    <div
      class="border border-green-500"
    >
      <div>
        Row 4
      </div>
    </div>
  </div>
</DocumentFragment>
`;
